simulator/ssh-transfer.go

```diff
@@ -0,0 +1,158 @@
+package simulator
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	simssh "github.com/alphasoc/flightsim/simulator/ssh"
+
+	bytesize "github.com/inhies/go-bytesize"
+)
+
+const defaultSendSize = 100 * bytesize.MB
+
+// TODO where will the server side live?
+var defaultTargetHosts = []string{"127.0.0.1:22", "127.0.0.1:9999"}
+
+// SSHTransfer defines this simulation.
+type SSHTransfer struct {
+	src      net.IP // Connect from this IP.
+	sendSize bytesize.ByteSize
+}
+
+// NewSSHTransfer creates a new SSH/SFTP simulator.
+func NewSSHTransfer() *SSHTransfer {
+	return &SSHTransfer{}
+}
+
+// Init sets the source IP for this simulation.
+func (s *SSHTransfer) Init(src net.IP) error {
+	s.src = src
+	return nil
+}
+
+// writeRandom writes sendSize bytes of 'random' data to the server.  An error is returned.
+func writeRandom(c *simssh.Client, handle string, sendSize bytesize.ByteSize) error {
+	// 8K writes.
+	const buffSize uint64 = 8192
+	toSendStr := sendSize.Format("%.0f", "B", false)
+	toSend, err := strconv.ParseUint(toSendStr[:len(toSendStr)-1], 10, 64)
+	if err != nil {
+		return fmt.Errorf("failed write: %v", err)
+	}
+	// Seed once.
+	rand.Seed(time.Now().UnixNano())
+	var i uint64
+	var bytes []byte
+	var payloadSize uint64
+	for i = 0; toSend > 0; i++ {
+		if toSend >= buffSize {
+			payloadSize = buffSize
+		} else {
+			payloadSize = toSend
+		}
+		bytes = make([]byte, payloadSize)
+		rand.Read(bytes)
+		// Don't care about the actual response here.  Just the error code.
+		_, err := c.SendWrite(handle, i*buffSize, &bytes)
+		if err != nil {
+			return fmt.Errorf("failed transfer: %v", err)
+		}
+		toSend -= payloadSize
+	}
+	return nil
+}
+
+// Simulate an ssh/sftp file transfer.
+func (s *SSHTransfer) Simulate(ctx context.Context, dst string) error {
+	// Auth.
+	signer, err := simssh.NewSignerFromKey()
+	if err != nil {
+		return err
+	}
+	// Create a Client that's ready to use for SSH/SFTP transfers.
+	c, err := simssh.NewClient("alphasoc", s.src, dst, signer)
+	if err != nil {
+		return err
+	}
+	defer c.Teardown()
```

Comment:
This seems fine but I think it's possible we open the connection in NewClient and still could return error. In such a case we wouldn't close it.

Reply:
Hmmm... did I miss closing something in client.go::NewClient()?

If sshClient.NewSession fails, it closes out the sshClient, which is supposed to close the underlying network connection nConn.  If any of the pipe creates fail, I close out both the sshClient and the session.

Where do you see the possible hole?

Reply:
You're right, I looked at the line 62 in NewClient and assumed nConn can leak and then skipped rest of the logic. Let me move the conversation in there.